src/messaging/ExchangeMessageDispatch.cpp

```diff
@@ -51,45 +51,60 @@ CHIP_ERROR ExchangeMessageDispatch::SendMessage(SessionManager * sessionManager,
     PayloadHeader payloadHeader;
     payloadHeader.SetExchangeID(exchangeId).SetMessageType(protocol, type).SetInitiator(isInitiator);
 
-    // If there is a pending acknowledgment piggyback it on this message.
-    if (reliableMessageContext->HasPiggybackAckPending())
+    if (session->RequireTCP())
```

Comment:
I don't understand why all the surgery here.  `reliableMessageContext->AutoRequestAck()` should be returning the right thing for TCP, and then the existing code would just work, no?

In any case, there shouldn't be separate cases for handling BTP and TCP here; they should take the same codepath.

Reply:
I hadn't paid close attention to that, but I think we can use it. Currently, it wouldn't work though as AutoRequestAck is set for all non-group sessions I think. So, if we set it for only UDP based sessions, then the existing logic should work for TCP.

Reply:
There is also the HasPiggybackAckPending() check. There are all these checks that are only relevant for MRP and should be avoided entirely for TCP messages instead of making TCP messages pass through them(even if false) I think.

Reply:
> as AutoRequestAck is set for all non-group sessions I think

OK, but then how do things work for BTP right now?  The answer is presumably that either they are buggy or they are already gated on the misnamed `RequireMRP()` thing.

But again: TCP is not special.  BTP and TCP should take the same codepaths.

I am fine with making that codepath clearer by avoiding calling functions, even if those functions would be guaranteed no-ops because they do their own checks.  But that should not be conditioned on "tcp".  It should be conditioned on whether the session allows MRP.

Reply:
So, it seems AutoRequestAck is set in the ExchangeContext ctor. Do we currently use MRP over BLE? We should not, right? 
>It should be conditioned on whether the session allows MRP.

Yes, that is how it is in the latest change.

Reply:
> Do we currently use MRP over BLE?

We shouldn't, and I am pretty sure we do not.

Reply:
That is what I would expect too. But it is set for BLE message in the ExchangeContext ctor. So, somewhere down the path, I guess it gets overridden by the RequireMRP() check, which is false for BLE.